### PR TITLE
Feat/rollback

### PR DESCRIPTION
**Pull request by sh-cha**
- rollup sync from initia-l1, celestia

### CODE REVIEW

File: rollupsync/provider/celestia.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+				err := cp.fetchBlock(ctx, searchMap)
+				if err != nil {
+					continue LOOP
+				}
+				if height > endHeight {
+					return nil
+				}
+				page = 1
+				cp.batchCh <- rstypes.BatchInfo{
+					L1QueryHeight: height - 1,
```
Comment (beer-1) [inline rollupsync/provider/celestia.go]: how about `BatchChainHeight`?

File: config/toml.go [hunk 2]
```diff
@@ -453,6 +453,21 @@ chunk_fetchers = "{{ .StateSync.ChunkFetchers }}"
...
+#######################################################
+[rollupsync]
+enable = {{ .RollupSync.Enable }}
+
+bridge_id = {{ .RollupSync.BridgeId }}
+
+# supported batch chain: l1, celestia
+batch_chain = "{{ .RollupSync.BatchChain }}"
+batch_submitter = "{{ .RollupSync.BatchSubmitter }}"
+
```
Comment (beer-1) [inline config/toml.go]:
```toml
// `batch_rpc` can be empty, if the `batch_chain` is "l1"
```

File: config/toml.go [hunk 3]
```diff
@@ -453,6 +453,21 @@ chunk_fetchers = "{{ .StateSync.ChunkFetchers }}"
...
+###         Rollup Sync Configuration Options       ###
+#######################################################
+[rollupsync]
+enable = {{ .RollupSync.Enable }}
+
+bridge_id = {{ .RollupSync.BridgeId }}
+
+# supported batch chain: l1, celestia
+batch_chain = "{{ .RollupSync.BatchChain }}"
+batch_submitter = "{{ .RollupSync.BatchSubmitter }}"
```
Comment (beer-1) [inline config/toml.go]: `inita1~` or `celestia1~` address depends on batch_chain config.

File: rollupsync/sync.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+func (rs *RollupSyncer) Start(ctx context.Context) (sm.State, error) {
+	targetBlockHeight, err := rs.outputProvider.GetLatestFinalizedBlock(ctx)
+	if err != nil {
+		return sm.State{}, err
+	}
+	rs.targetBlockHeight = targetBlockHeight
+	if rs.state.LastBlockHeight >= int64(targetBlockHeight) {
+		return rs.state, err
+	}
+	start, _ := rs.blockExec.Store().GetRollupSyncL1Block()
```
Comment (beer-1) [inline rollupsync/sync.go]: `GetRollupSyncBatchChainHeight()`?

Comment (beer-1) [reply rollupsync/sync.go]: is it okay to ignore error?

File: rollupsync/provider/utils.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+	rpchttp "github.com/cometbft/cometbft/rpc/client/http"
+
+	celblob "github.com/celestiaorg/go-square/blob"
+)
+
+var (
+	DEFAULT_FETCH_INTERVAL  = 10 // millisecond
+	DEFAULT_TXS_PER_PAGE    = 100
+	DEFAULT_HEIGHT_INTERVAL = 100000
```
Comment (beer-1) [inline rollupsync/provider/utils.go]: Seems good to move to config

File: rollupsync/const.go [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package rollupsync
+
+// https://github.com/initia-labs/OPinit/blob/feat/celestia/bots/src/worker/batchSubmitter/batchSubmitter.ts#L125C7-L125C40
+const MAX_BATCH_BYTES = 500_000
+
+// temporary
+const MAX_BATCH_CHUNK = 1
```
Comment (beer-1) [inline rollupsync/const.go]: This seems too small, so good to use 10 or bigger number or make this also configurable

File: rollupsync/sync.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+			batch = append(batch, batchInfo.Batch...)
+			rawData, err := DecompressBatch(batch)
+
+			// in the case of failure, a single batch might be separated several chunks of bytes
+			// TODO: calculate maximum batch chunks and restrict appendending chunks continuously
+			// to avoid a situation that the raw batch is abnormal.
+			if err != nil {
+				if chunks >= MAX_BATCH_CHUNK {
+					rs.logger.Error("error decompressing batch", "error", err)
+					return
```
Comment (beer-1) [inline rollupsync/sync.go]: maybe we need some way to stop this rollup sync process when this go thread met error by ctx <- done?  or quite

File: rollupsync/sync.go [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+			dataLength := len(rawData)
+			rawBlocks := rawData[:dataLength-1]
+			rawCommit := rawData[dataLength-1]
+
+			for _, blockBytes := range rawBlocks {
+				pbb := new(cmtproto.Block)
+				err := proto.Unmarshal(blockBytes, pbb)
+				if err != nil {
+					rs.logger.Error("error reading block", "error", err)
+					return
```
Comment (beer-1) [inline rollupsync/sync.go]: This too

File: rollupsync/sync.go [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+				}
+
+				block, err := types.BlockFromProto(pbb)
+				if err != nil {
+					rs.logger.Error("error from proto block", "error", err)
+					return
```
Comment (beer-1) [inline rollupsync/sync.go]: here too

File: rollupsync/sync.go [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+				}
+				rs.blockCh <- rstypes.BlockInfo{
+					Block: block,
+				}
+			}
+
+			pbc := new(cmtproto.Commit)
+			err = proto.Unmarshal(rawCommit, pbc)
+			if err != nil {
+				panic(fmt.Sprintf("error reading block seen commit: %v", err))
```
Comment (beer-1) [inline rollupsync/sync.go]: panic to stop

File: rollupsync/sync.go [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+			}
+
+			commit, err := types.CommitFromProto(pbc)
+			if err != nil {
+				panic(fmt.Errorf("converting seen commit: %w", err))
```
Comment (beer-1) [inline rollupsync/sync.go]: panic to stop

File: rollupsync/sync.go [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+		case <-ctx.Done():
+			rs.batchProvider.Quit()
+			return state, ctx.Err()
+		case blockInfo := <-rs.blockCh:
+			block := blockInfo.Block
+			if block != nil {
+				if state.LastBlockHeight+1 > block.Height {
+					continue LOOP
+				} else if state.LastBlockHeight+1 < block.Height {
+					panic(fmt.Errorf("sync height mismatch; expected %d, got %d", state.LastBlockHeight+1, block.Height))
```
Comment (beer-1) [inline rollupsync/sync.go]: maybe good to stop with error not panic.

File: rollupsync/sync.go [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+				}
+				blockPartSetHeader := blockParts.Header()
+				blockID := types.BlockID{Hash: block.Hash(), PartSetHeader: blockPartSetHeader}
+
+				// seen commit saves at next block
+				rs.store.SaveBlock(block, blockParts, nil)
+
+				state, err = rs.blockExec.ApplyBlock(state, blockID, block)
+				if err != nil {
+					panic(fmt.Sprintf("Failed to process committed block (%d:%X): %v", block.Height, block.Hash(), err))
```
Comment (beer-1) [inline rollupsync/sync.go]: This too

File: rollupsync/sync.go [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+				}
+				if block.Height == int64(rs.targetBlockHeight) {
+					rs.batchProvider.Quit()
+					break LOOP
+				}
+			} else {
+				rs.blockExec.Store().SetRollupSyncL1Block(blockInfo.L1QueryHeight)
```
Comment (beer-1) [inline rollupsync/sync.go]: BatchChainHeight

File: rollupsync/types/channel.go [hunk 15]
```diff
@@ -0,0 +1,13 @@
+package types
+
+import "github.com/cometbft/cometbft/types"
+
+type BatchInfo struct {
+	L1QueryHeight int64
```
Comment (beer-1) [inline rollupsync/types/channel.go]: BatchChainHeight

File: rollupsync/types/channel.go [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+	Batch         []byte
+}
+
+type BlockInfo struct {
+	L1QueryHeight int64
```
Comment (beer-1) [inline rollupsync/types/channel.go]: BatchChainHeight

File: rollupsync/sync.go [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+		for batchInfo := range batchCh {
+			if batchInfo.Batch == nil {
+				rs.blockCh <- rstypes.BlockInfo{
+					L1QueryHeight: batchInfo.L1QueryHeight,
+				}
+				continue
+			}
+			rs.logger.Debug("Receive a batch chunk")
+
+			chunks++
```
Comment (beer-1) [inline rollupsync/sync.go:176]: chunks should be reset when decompression succeed

File: rpc/core/types/responses.go [hunk 18]
```diff
@@ -265,3 +265,8 @@ type ResultInvalidBlock struct {
 type ResultBlockBulk struct {
 	Blocks [][]byte `json:"blocks"`
 }
+
+// Result of block bulk
```
Comment (beer-1) [inline rpc/core/types/responses.go]: raw commit bytes

File: state/store.go [hunk 19]
```diff
@@ -180,6 +188,25 @@ func (store dbStore) loadState(key []byte) (state State, err error) {
 	return *sm, nil
 }
 
+// it is to save current validators when executor is changed
```
Comment (beer-1) [inline state/store.go]: // overwrite validator set if the executor is changed

File: rollupsync/sync.go [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+					rs.batchProvider.Quit()
+					rs.logger.Error("failed to make ",
+						"height", block.Height,
+						"err", err.Error())
+					return state, err
+				}
+				blockPartSetHeader := blockParts.Header()
+				blockID := types.BlockID{Hash: block.Hash(), PartSetHeader: blockPartSetHeader}
+
+				// seen commit saves at next block
```
Comment (beer-1) [inline rollupsync/sync.go]: don't need to save seen commit here, seen commit is used only in consensus.

File: state/execution.go [hunk 21]
```diff
@@ -553,6 +563,16 @@ func BuildExtendedCommitInfo(ec *types.ExtendedCommit, valSet *types.ValidatorSe
 	}
 }
 
+func checkExecutorChanged(abciUpdates []abci.ValidatorUpdate) bool {
+	for i := range abciUpdates {
```
Comment (beer-1) [inline state/execution.go]: old validators can be multiple, so maybe we need to handle that case here.

**Review by beer-1 — COMMENTED** [review]
LGTM 👍, seems need minor changes to make const to configurable.

need change in `checkExecutorChanged`

File: rollupsync/sync.go (hunk 4 above)
Comment (sh-cha) [reply rollupsync/sync.go]: it is to start from scratch if there is no saved height

File: config/config.go [hunk 22]
```diff
@@ -952,6 +954,52 @@ func (cfg *BlockSyncConfig) ValidateBasic() error {
...
+	return DefaultRollupSyncConfig()
+}
+
+// ValidateBasic performs basic validation.
+func (cfg *RollupSyncConfig) ValidateBasic() error {
+	if cfg.Enable {
+		if cfg.BridgeId == 0 {
+			return errors.New("bridge id is required")
+		}
+		if cfg.BatchChain == "l1" || cfg.BatchChain == "celestia" {
```
Comment (beer-1) [inline config/config.go]: seems condition is wrong

File: config/config.go [hunk 23]
```diff
@@ -952,6 +954,52 @@ func (cfg *BlockSyncConfig) ValidateBasic() error {
...
+	}
+}
+
+// TestBlockSyncConfig returns a default configuration for the block sync.
+func TestRollupSyncConfig() *RollupSyncConfig {
+	return DefaultRollupSyncConfig()
+}
+
+// ValidateBasic performs basic validation.
+func (cfg *RollupSyncConfig) ValidateBasic() error {
```
Comment (beer-1) [inline config/config.go:1005]: This function should be called in L132

File: rollupsync/utils.go [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+	"io"
+
+	"github.com/cometbft/cometbft/rollupsync/types"
+)
+
+func DecompressBatch(b []byte) ([][]byte, error) {
+	br := bytes.NewReader(b)
+	r, err := gzip.NewReader(br)
+	if err != nil {
+		fmt.Println("1", err)
```
Comment (beer-1) [inline rollupsync/utils.go]: debug print

File: rollupsync/utils.go [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+		return nil, err
+	}
+	defer r.Close()
+
+	res, err := io.ReadAll(r)
+	if err != nil {
+		fmt.Println("2", err)
```
Comment (beer-1) [inline rollupsync/utils.go]: debug print

**Review by beer-1 — APPROVED** [review]
https://github.com/initia-labs/cometbft/pull/2/files

have made some cleanup pr, so good to merge after merge the one

**Pull request merged**